platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/merkle/AbstractHashListener.java

```diff
@@ -131,21 +131,36 @@ public synchronized void onLeafHashed(final VirtualLeafRecord<K, V> leaf) {
     }
 
     @Override
-    public synchronized void onHashingCompleted() {
-        if (!nodes.isEmpty() || !leaves.isEmpty()) {
+    public void onHashingCompleted() {
+        boolean needFinalFlush = false;
+        synchronized (this) {
+            if (!nodes.isEmpty() || !leaves.isEmpty()) {
+                needFinalFlush = true;
+            }
+        }
+        if (needFinalFlush) {
             flush(nodes, leaves);
         }
         nodes = null;
         leaves = null;
```

Comment:
Since elsewhere in this class we always assign values to these fields under the `this` lock, I suggest to modify the implementation as follows:

```suggestion
        final List<VirtualLeafRecord<K, V>> leaves = this.leaves;
        final List<VirtualHashRecord> nodes = this.nodes;
        synchronized (this) {
            nodes = null;
            leaves = null;
        }
        if (!nodes.isEmpty() || !leaves.isEmpty()) {
            flush(nodes, leaves);
        }
```

Reply:
OK, I will fix this. Note that this.nodes and this.leaves should be accessed and modified in a synchronized block.